src/tools/ceph_objectstore_tool.cc

```diff
@@ -680,59 +685,86 @@ int do_trim_pg_log(ObjectStore *store, const coll_t &coll,
 	continue;
       if (p->key().substr(0, 7) == string("missing"))
 	continue;
-      if (p->key().substr(0, 4) == string("dup_"))
-	continue;
-
       bufferlist bl = p->value();
       auto bp = bl.cbegin();
-      pg_log_entry_t e;
-      try {
-	e.decode_with_checksum(bp);
-      } catch (const buffer::error &e) {
-	cerr << "Error reading pg log entry: " << e.what() << std::endl;
-      }
-      if (debug) {
-	cerr << "read entry " << e << std::endl;
+
+      if (p->key().substr(0, 4) == string("dup_")) {  
+        pg_log_dup_t dup;
+	try {
+	  dup.decode(bp);
+	  pg_log.dups.push_back(dup);
+	  if (debug) {
+            cerr << "read entry " << dup << std::endl;
+          }
+	} catch (buffer::error& e) {
+	  cerr << "error reading log dup: " << p->key() << std::endl;
+	  return -EFAULT;
+	}
+      } else {
+        pg_log_entry_t e;
+        try {
+          e.decode_with_checksum(bp);
+          pg_log.log.push_back(e);
+	  if (debug) {
+            cerr << "read entry " << e << std::endl;
+          }
+	} catch (const buffer::error &e) {
+          cerr << "Error reading pg log entry: " << e.what() << std::endl;
+        }
       }
-      if (e.version.version > trim_to) {
-	done = true;
-	break;
+      // We have enough pg logs entries to trim
+      if ((pg_log.log.size() + pg_log.dups.size()) >= trim_at_once) {
+        done = true;
```

Comment:
I don't think we want to set `done` here. I think we want `pg_log.trim` to be called multiple times, being reset after processing a chunk, unless all the keys are processed.

Alternatively, we could not break here, collecting instead all potential entries to trim into `pg_log`. Then call `pg_log.trim` only once, obtaining the full lists of the trimmed keys, and then call `t.omap_rmkeys` multiple times in `trim_at_once` chunks. The problem is though that in this scenario `pg_log.trim` may build very large `trimmed` and `trimmed_dups` lists and this seems that we may not allow.